pgrx/src/nullable.rs

```diff
@@ -0,0 +1,523 @@
+    fn next_valid_idx(&self, idx: usize) -> Option<usize> {
+        // Next elem (one after this) would be past the end
+        // of the container
+        if (idx + 1) >= self.0.len() {
+            return None;
+        }
+        let mut resulting_idx = 0;
+        for bit in &(*self.0)[(idx + 1)..] {
+            // Postgres nullbitmaps are 1 for "valid" and 0 for "null"
+            resulting_idx += (*bit) as usize;
+        }
+        Some(resulting_idx)
+    }
+}
+
+pub struct BoolSliceNulls<'a>(pub &'a [bool]);
+
+impl<'a> NullLayout<usize> for BoolSliceNulls<'a> {
+    fn len(&self) -> usize {
+        <[bool]>::len(&self.0)
+    }
+
+    fn has_nulls(&self) -> bool {
+        let mut has_null = false;
+        for value in self.0 {
+            if *value {
+                has_null = true;
+            }
+        }
+        has_null
+    }
+    // In a Postgres bool-slice implementation of a null layout,
+    // 1 is "null", 0 is "valid"
+    fn is_valid(&self, idx: usize) -> Option<bool> {
+        if idx < self.0.len() {
+            // Invert from 0 being valid to 1 being valid.
+            Some(!self.0[idx])
+        } else {
+            None
+        }
+    }
+
+    fn count_nulls(&self) -> usize {
+        let mut count = 0;
+        for elem in self.0 {
+            count += *elem as usize;
+        }
+        count
+    }
+}
+
+// Postgres arrays using a bool array to map which values are null and which
+// values are valid will always be contiguous - that is, the underlying data
+// buffer will actually be big enough to contain layout.len() slots of type T
+// (give or take padding)
+impl<'a> ContiguousNullLayout<usize> for BoolSliceNulls<'a> {}
+
+/// Strict i.e. no nulls.
+/// Useful for using nullable primitives on non-null structures,
+/// especially when this needs to be determined at runtime.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Copy, Clone)]
+pub struct StrictNulls(pub usize);
+
+impl NullLayout<usize> for StrictNulls {
+    fn len(&self) -> usize {
+        self.0
+    }
+
+    fn has_nulls(&self) -> bool {
+        false
+    }
+
+    fn is_valid(&self, idx: usize) -> Option<bool> {
+        (idx < self.0).then_some(true)
+    }
+
+    fn is_null(&self, idx: usize) -> Option<bool> {
+        (idx < self.0).then_some(false)
+    }
+
+    fn count_nulls(&self) -> usize {
+        0
+    }
+}
+// No skipping when there are no nulls.
+impl ContiguousNullLayout<usize> for StrictNulls {}
+
+// Intended to represent any one of the three nullable container layouts used
+// by Postgres.
+pub enum AnyNullLayout<'a> {
+    /// Bitmap of where null slots are in this container.
+    /// For example, 00001001 would represent:
+    /// \[value, value, value, value, null, value value, null\]
+    /// However, the underlying data buffer would be:
+    /// \[value, value, value, value, value, value\]
+    /// because of the skip behavior
+    Bitmap(BitSliceNulls<'a>),
```

Comment:
other way around, actually, `[valid, valid, valid, valid, null, valid, valid, null]` is `0b11110110`. you implement this correctly elsewhere, this is just a doc bug.

Reply:
Thanks for the heads-up!